pkg/controller/serviceaccount_secret_refresher/serviceaccount_secret_refresher.go

```diff
@@ -0,0 +1,190 @@
+		return fmt.Errorf("failed to construct controller: %w", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &corev1.ServiceAccount{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to construct watch for ServiceAccounts: %w", err)
+	}
+	if err := c.Watch(&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(secretMapper)); err != nil {
+		return fmt.Errorf("failed to construct watch for Secrets: %w", err)
+	}
+
+	return nil
+}
+
+func secretMapper(o ctrlruntimeclient.Object) []reconcile.Request {
+	secret, ok := o.(*corev1.Secret)
+	if !ok {
+		logrus.WithField("type", fmt.Sprintf("%T", o)).Error("Got an object that was not a secret")
+		return nil
+	}
+
+	sa, ok := secret.Annotations[corev1.ServiceAccountNameKey]
+	if !ok {
+		return nil
+	}
+
+	return []reconcile.Request{{NamespacedName: types.NamespacedName{Namespace: secret.Namespace, Name: sa}}}
+}
+
+type reconciler struct {
+	client           ctrlruntimeclient.Client
+	filter           func(reconcile.Request) bool
+	log              *logrus.Entry
+	removeOldSecrets bool
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	l := r.log.WithField("request", req.String())
+	res, err := r.reconcile(ctx, l, req)
+	if err != nil {
+		l.WithError(err).Error("Reconciliation failed")
+	}
+	if res == nil {
+		res = &reconcile.Result{}
+	}
+	return *res, err
+}
+
+func (r *reconciler) reconcile(ctx context.Context, l *logrus.Entry, req reconcile.Request) (*reconcile.Result, error) {
+	if !r.filter(req) {
+		return nil, nil
+	}
+
+	sa := &corev1.ServiceAccount{}
+	if err := r.client.Get(ctx, req.NamespacedName, sa); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, fmt.Errorf("failed to get serviaccount %s: %w", req.String(), err)
+	}
+
+	var imagePullSecetsToKeep []corev1.LocalObjectReference
+	var requeueAfter time.Duration
+	for _, pullSecretRef := range sa.ImagePullSecrets {
+		creationTimestamp, err := r.creationTimestampSecret(ctx, req.Namespace, pullSecretRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check docker secrets creationTimestamp: %w", err)
+		}
+		if isObjectLessThan30DaysOld(*creationTimestamp) {
+			imagePullSecetsToKeep = append(imagePullSecetsToKeep, pullSecretRef)
+			if newReqeueAfter := -time.Since(creationTimestamp.Time.Add(30 * 24 * time.Hour)); requeueAfter == 0 || newReqeueAfter < requeueAfter {
+				requeueAfter = newReqeueAfter
+			}
+			continue
+		}
+		l.WithField("imagepullsecretname", pullSecretRef.Name).Info("Not keeping image pull secret")
+	}
+
+	var tokenSecretsToKeep []corev1.ObjectReference
+	for _, tokenSecretRef := range sa.Secrets {
+		creationTimestamp, err := r.creationTimestampSecret(ctx, req.Namespace, tokenSecretRef.Name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to check if secret should be kept: %w", err)
+		}
+		if isObjectLessThan30DaysOld(*creationTimestamp) {
```

Comment:
semi-nit: should 30 be a param or a const to avoid hardcoding? What if we nee

Reply:
@alvaroaleman Let's also rename the method so it does not hardcode the number? So that if we change the number we don't need the method name. Perhaps `isCreatedRecently` or smt

Reply:
@petr-muller updated